core/trino-main/src/test/java/io/trino/server/security/TestResourceSecurity.java

```diff
@@ -190,13 +202,66 @@ public void testPasswordAuthenticator()
                         .build())
                 .build()) {
             server.getInstance(Key.get(PasswordAuthenticatorManager.class)).setAuthenticator(TestResourceSecurity::authenticate);
-            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl());
+            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl(false));
             HttpServerInfo httpServerInfo = server.getInstance(Key.get(HttpServerInfo.class));
             assertAuthenticationDisabled(httpServerInfo.getHttpUri());
             assertPasswordAuthentication(httpServerInfo.getHttpsUri());
         }
     }
 
+    @Test
+    public void testPasswordAuthenticatorUserMapping()
+            throws Exception
+    {
+        try (TestingTrinoServer server = TestingTrinoServer.builder()
+                .setProperties(ImmutableMap.<String, String>builder()
+                        .putAll(SECURE_PROPERTIES)
+                        .put("http-server.authentication.type", "password")
+                        .put("http-server.authentication.password.user-mapping.pattern", ALLOWED_USER_MAPPING_PATTERN)
+                        .build())
+                .setAdditionalModule(binder -> jaxrsBinder(binder).bind(TestResource.class))
+                .build()) {
+            server.getInstance(Key.get(PasswordAuthenticatorManager.class)).setAuthenticator(TestResourceSecurity::authenticate);
+            server.getInstance(Key.get(AccessControlManager.class)).addSystemAccessControl(new TestSystemAccessControl(false));
+            HttpServerInfo httpServerInfo = server.getInstance(Key.get(HttpServerInfo.class));
+
+            // Test sets basic auth user and X-Trino-User, and the authenticator is performing user mapping.
+            // Normally this would result in an impersonation check to the X-Trino-User, but the password
+            // authenticator has a hack to clear X-Trino-User in this case.
+            Request request = new Request.Builder()
+                    .url(getLocation(httpServerInfo.getHttpsUri(), "/username"))
+                    .addHeader("Authorization", Credentials.basic(TEST_USER_LOGIN, TEST_PASSWORD))
+                    .addHeader("X-Trino-User", TEST_USER_LOGIN)
```

Comment:
In this test, add a check what's the effective session user on the server. (You can run eg. `SELECT current_user`)
I see you check response header "user", but as we modify headers within filters, this is not sufficient from integration test purposes.

(If the framework of this test doesn't allow you to run queries, you can easily test this with a product test)

Reply:
This is exactly what this test does.  I am adding a custom JAXRS resouce `TestResource`, and this resource calls `extractAuthorizedIdentity` which is the code we use in all of our secure REST endpoints to get the current user's identity.  This uses the same code as the query submission endpoint.  The test resource then takes that username from the Identity, and put it into a custom response header for this test named `user` (not all Trino request and response headers start with `X-Trino-`), and then I check the value of that response header below.

Reply:
> TestResource, and this resource calls extractAuthorizedIdentity which is the code we use in all of our secure REST endpoints to get the current user's identity.

`extractAuthorizedIdentity` is similar (or equal) code as query submission, but it's not the same code.

query submission doesn't go thru `extractAuthorizedIdentity`

Reply:
I agree with @findepi, even though now the code looks the same, it might not be the same in the future so, the test should test `QuestSessionSupplier#createSession` rather than `HttpRequestSessionContext#extractAuthorizedIdentity`.